### ISSUE TITLE
[css-cascade-5] Add a R/W `CSSStyleSheet.layer` attribute for the layer name

### ISSUE THREAD

**Issue by xiaochengh** [w3c/csswg-drafts#7002]
I'm proposing extending the `CSSStyleSheet` interface with a new attribute, `layer`.

On getting, it should return the layer name if the sheet is entirely in some cascade layer. In particular:
- If this is an `@import`-ed stylesheet, return the `layerName` attribute value of the `CSSImportRule`
- If this stylesheet is added via a `<link>` element (depending on https://github.com/whatwg/html/issues/7540), then return the name of that layer
- If this is a constructed stylesheet with a layer name previous set by JS (see below), return that layer name
- Otherwise, return `null`

On setting:
- If this is a constructed stylesheet and the new value is a non-null string, then the attribute value can be modified, and the stylesheet is treated as entirely in the given layer
- Otherwise, returns `null`

The primary use case is to allow adding layered constructed stylesheets, especially if the stylesheet is a 3rd party CSS module, and we want to use that to set up a web component (which we currently don't have a good way to make it layered):

```javascript
import sheet from 3p-foo.css;

sheet.layer = 'new-layer';
shadowRoot.adoptedStylesheets = [sheet];
```

**Comment by tabatkins** [comment]
Sounds good to me; this is an obvious hole currently both on the reading side (can't tell from inspection if a stylesheet was imported or linked with a layer) and on the writing side (can't have constructed sheets with the equivalent functionality of importing/linking into a particular layer).

**Comment by fantasai** [comment]
How should anonymous layers work? Maybe represent them with the empty string?

**Comment by xiaochengh** [comment]
> How should anonymous layers work? Maybe represent them with the empty string?

Yes. This is also consistent with how `CSSImportRule.layerName` distinguishes importing into an anonymous layer (empty string) vs. not importing into any layer (null).

**Comment by tabatkins** [comment]
Ah, we should match that name, too, and use `.layerName` on CSSStyleSheet.

**Comment by tabatkins** [comment]
(Or switch them both to `.layer`.)

**Comment by xiaochengh** [comment]
If it's a readonly attribute, how do we support the JS import syntax (which is the major use case I'd like to support)? We don't explicitly call its constructor so we can't pass the layer name in that way.

> &lt;TabAtkins&gt; emilio: When you use `import sheet from style.css`, can you set the layer there?
> &lt;TabAtkins&gt; emilio: I think in JS you have a trailing object that lets you set the mimetype?
> &lt;TabAtkins&gt; TabAtkins: Yes, but I'm unsure whether that's meant to be just assertions or can have extra data

@tabatkins When you said "yes", were you talking about setting the layer or the mimetype?

**Comment by tabatkins** [comment]
"yes" to trailing object; I just wasn't sure if it's meant to be able to carry additional data beyond the mime type.

**Comment by xiaochengh** [comment]
Then the resolution is inadequate. We still need a way to make it work on the JS-imported style modules. The easiest way that I can see is still to make it writable on constructed stylesheets.

**Comment by emilio** [comment]
@xiaochengh can't the trailing object carry the layer name? If not, why not?

**Comment by xiaochengh** [comment]
Sorry I misunderstood. Didn't realize you were talking about import assertion. Yeah, we can pass a layer name there.

There's also one thing that I think we missed in the previous discussion. The same sheet can be imported for multiple times into different layers (e.g., when used by different modules), which means we need to make sure that each import creates a different `CSSStyleSheet` object. Is this already guaranteed, or do we need more work?

**Comment by mirisuzanne** [comment]
Agenda+ to get @xiaochengh's question answered: 

> The same sheet can be imported for multiple times into different layers (e.g., when used by different modules), which means we need to make sure that each import creates a different CSSStyleSheet object. Is this already guaranteed, or do we need more work?

@fantasai and I are not clear on the answer.

**Comment by fantasai** [comment]
If CSSStyleSheet instances are not unique to their usage, and one can be imported in multiple places, then we can't have a layerName attribute. Instead, the importing mechanism needs to accept/track the layer name.

**Comment by chrishtr** [comment]
@dandclark do you know the answer to @xiaochengh's question?

**Comment by zcorpan** [comment]
CSS module scripts are defined here https://html.spec.whatwg.org/multipage/webappapis.html#creating-a-css-module-script

which is called from https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-single-module-script

**Comment by dbaron** [comment]
Based on my memory from the CSS Modules TAG review (and a quick re-read to refresh my memory) in w3ctag/design-reviews#405, part of the idea of CSS modules is definitely that they are a new mechanism that allows the same stylesheet object to be referenced from multiple places, and then be used by adding to `adoptedStyleSheets` in multiple places.  But it's not clear to me that there's any way to use a CSS module in a way that involves a layer name.

However, there was an issue related to `@import` (which I think is currently forbidden from CSS Modules) that was being deferred at the time of that TAG review.  Depending on how that issue is resolved, it maybe be possible to use [CSS `@import` syntax](https://drafts.csswg.org/css-cascade-5/#at-import) to extend the module graph, and thus create layer-name-specific imports that refer to sheets in the module graph.  I'm not sure what eventually happened with that issue.

**Comment by dandclark** [comment]
Separate imports of a CSSStyleSheet from a given URL will share the same stylesheet instance. This is analogous to how multiple imports of a given JS module reuse the same module.

That seems to be problematic for `layerName`; we'd need a way to guarantee that importing with a new layerName triggers a new instance of the sheet to be created, which isn't currently something supported by import assertions.

**Comment by dandclark** [comment]
There's been discussion in the past about extending the `import` syntax to allow attributes that change things about the imported object (which is not allowed in the `assert` clause). This would create a new object for each `import` with different properties:
```js
import stylesA from "./style1.css" assert { type: "css"} with {layerName: "foo"}; // Creates stylesheet instance
import stylesB from "./style1.css" assert { type: "css"} with {layerName: "foo"}; // Same instance as stylesA
import stylesC from "./style1.css" assert { type: "css"} with {layerName: "bar"}; // New instance because value in `with` clause is different
```
But this would require a syntax change going through TC39.
